File: Data Structures & Algorithms/two-integer-sum-ii/submission-0.py
# Time Complexity: O(n), two pointer approach, single pass through list
# Space Complexity: O(1), only use constant space for variables

class Solution:
    def twoSum(self, numbers: List[int], target: int) -> List[int]:
        answer = []
        left = 0
        right = len(numbers) - 1
        while left < right:
            sum = numbers[left] + numbers[right]
            if sum < target:
                left += 1
            elif sum > target:
                right -= 1
            else:
                break
        answer.append(left + 1)
        answer.append(right + 1)
        return answer

File: Data Structures & Algorithms/max-water-container/submission-0.java
// Time Complexity: O(n), two pointer approach, single pass through array
// Space Complexity: O(1), only use constant space for variables

class Solution {
    public int maxArea(int[] heights) {
        // Area = Math.min(left height, right heigh) * width
        // Always shift the shorter height of the two bars -> Explores all meaningful area
        int left = 0, right = heights.length - 1;
        int maxArea = Integer.MIN_VALUE;

        while (left < right) {
            int currArea = Math.min(heights[left], heights[right]) * (right - left);
            maxArea = Math.max(maxArea, currArea);

            if (heights[left] < heights[right]) {
                left++;
            } else {
                right--;
            }
        }

        return maxArea;
    }
}
